oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/AsyncCheckpointCreator.java

```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Override
+    public void run() {
+        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+        String creationTimeStamp = String.valueOf(cal.getTimeInMillis());
+        String creationTimeISOFormat = ISO8601.format(cal);
+        String checkpoint = store.checkpoint(checkpointLifetimeInSeconds * 1000, Map.of(
+                CHECKPOINT_CREATOR_KEY, AsyncCheckpointCreator.class.getSimpleName(),
+                CHECKPOINT_CREATED_KEY, creationTimeISOFormat,
+                CHECKPOINT_CREATED_TIMESTAMP_KEY, creationTimeStamp,
+                CHECKPOINT_THREAD_KEY, Thread.currentThread().getName(),
+                CHECKPOINT_NAME_KEY, name));
+        log.info("[{}] Created checkpoint {} with creation time {}", name, checkpoint, creationTimeISOFormat);
+
+        // Get a list of checkpoints filtered on the basis of CHECKPOINT_NAME_KEY (name). This is done using the
+        // getFilteredCheckpoints in the IndexUtils, which gets all the checkpoints in the node store and then filters the list based on
+        // the provided predicate using the checkpoint info map associated with the checkpoints.
+        // The filtered checkpoints list is then sorted based on the CHECKPOINT_CREATED_TIMESTAMP_KEY (created-epoch).
+        // Both the initial filtering and sorting is done based on the information from the associated checkpoint info map of a given checkpoint.
+        Set<String> sortedCheckpointSet = IndexUtils.getSortedCheckpointMap(IndexUtils.getFilteredCheckpoints(store,
+                entry -> name.equals(entry.getValue().get(CHECKPOINT_NAME_KEY))), CHECKPOINT_CREATED_TIMESTAMP_KEY).keySet();
+        int counter = sortedCheckpointSet.size();
+        // Iterate over the sortedCheckpointSet which is sorted based on the creation timestamp -> the oldest first
+        // We try and delete the checkpoints as long as the counter is greater than minConcurrentCheckpoints
+        // This ensures that the system always has concurrent checkpoints equal to or greater than the configured minConcurrentCheckpoints
+        for (String cp : sortedCheckpointSet) {
+            // Delete the checkpoint as long as the checkpoint count is greater than concurrentCheckpoints
+            if (counter > minConcurrentCheckpoints) {
+                if(store.release(cp) ) {
+                    log.info("[{}] Deleted checkpoint {}", name, cp);
+                } else {
+                    log.warn("[{}] Unable to delete checkpoint {}. Removal will be attempted again in next run.", name, cp);
+                }
+            } else {
+                break;
+            }
+            // Decrement the counter irrespective of the outcome of the checkpoint deletion.
+            // If we don't decrement the counter in case of a failure while trying to delete a checkpoint,
+            // the next iteration will try to delete a comparatively newer checkpoint and keep the older one in the system (which we don't one).
```

Comment:
```suggestion
            // the next iteration will try to delete a comparatively newer checkpoint and keep the older one in the system (which we don't want).
```
I'm not totally sure you mean "want".